File: go/CountAvgNode/CountAvgNode.go
package main

import "fmt"

type BinaryNode struct {
	Val   int
	Left  *BinaryNode
	Right *BinaryNode
}

// Time -> 0(N) Space -> 0(logN)
func countNodes(root *BinaryNode) int {
	count := 0
	recurse(root, &count)
	return count
}

func recurse(root *BinaryNode, count *int) (int, int) {
	if root == nil {
		return 0, 0
	}

	leftSum, leftSize := recurse(root.Left, count)
	rightSum, rightSize := recurse(root.Right, count)

	currSum := root.Val + leftSum + rightSum
	currSize := leftSize + rightSize + 1

	if (currSum / currSize) == root.Val {
		*count++
	}

	return currSum, currSize
}

func main() {
	leaf1 := &BinaryNode{Val: 0, Left: nil, Right: nil}
	leaf2 := &BinaryNode{Val: 1, Left: nil, Right: nil}
	leaf3 := &BinaryNode{Val: 6, Left: nil, Right: nil}
	mid1 := &BinaryNode{Val: 8, Left: leaf1, Right: leaf2}
	mid2 := &BinaryNode{Val: 5, Left: nil, Right: leaf3}
	root := BinaryNode{Val: 4, Left: mid1, Right: mid2}
	fmt.Println(countNodes(&root))
}

File: python/countAvgNode.py
class BinaryNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


# Time -> 0(N) Space -> 0(logN)
def countNodes(root: BinaryNode | None) -> int:
    count = [0]
    recurse(root, count)
    return count[0]


def recurse(root: BinaryNode | None, count: list[int]) -> tuple[int, int]:
    if root is None:
        return 0, 0

    leftSum, leftSize = recurse(root.left, count)
    rightSum, rightSize = recurse(root.right, count)

    currSum = leftSum + rightSum + root.val
    currSize = leftSize + rightSize + 1

    if (currSum // currSize) == root.val:
        count[0] += 1

    return currSum, currSize


leaf1 = BinaryNode(val=0, left=None, right=None)
leaf2 = BinaryNode(val=1, left=None, right=None)
leaf3 = BinaryNode(val=6, left=None, right=None)
mid1 = BinaryNode(val=8, left=leaf1, right=leaf2)
mid2 = BinaryNode(val=5, left=None, right=leaf3)
root = BinaryNode(val=4, left=mid1, right=mid2)

print(countNodes(root))
